### ISSUE TITLE
Problem routing with kubernetes and Deis 1.11

### ISSUE THREAD

**Issue by Grindizer** [deis/deis#4739]
Hi everyone, 

I have a really basic setup with 3 nodes running Deis 1.11 on AWS, the nodes are behind an ELB.
My setup is stateless with an RDS and an S3 bucket for the registry.
(I am able to deploy/access/scale an application when fleet is used as scheduler) 

when I install and use Kubernetes for scheduling, I am not able to route to my application anymore, actually I can see the kubernetes services created, an VIP is affected to the service, and I can see my Pods as well, but the virtual IP of the service is only known on the nodes that run at least one pod of my application.

Here a couple of command to explain a bit more, 
(inv is an inventory of my 3 nodes, controller is the kube controller node). 
When I deployed the application with "scale cmd=2"
```
ubuntu@ip-10-0-8-122:/tmp$ ansible -i inv controller -m shell -a "./kubectl --namespace=alfred-kube get pods -o wide"
10.0.124.137 | success | rc=0 >>
NAME                       READY     STATUS    RESTARTS   AGE       NODE
alfred-kube-v2-cmd-c8pad   1/1       Running   0          1m        10.0.124.137
alfred-kube-v2-cmd-d1uvz   1/1       Running   0          15h       10.0.150.224

ubuntu@ip-10-0-8-122:/tmp$ ansible -i inv controller -m shell -a "./kubectl --namespace=alfred-kube get services"
10.0.124.137 | success | rc=0 >>
NAME        LABELS             SELECTOR                    IP(S)           PORT(S)
app-15746   name=alfred-kube   name=alfred-kube,type=cmd   10.100.237.21   80/TCP

ubuntu@ip-10-0-8-122:/tmp$ ansible -i inv controller -m shell -a "./kubectl --namespace=alfred-kube describe service app-15746 | grep IP"
10.0.124.137 | success | rc=0 >>
Type:			ClusterIP
IP:			10.100.237.21
```
With this config a can only access my service on the 2 node that run my Pod:
```
ubuntu@ip-10-0-8-122:/tmp$ ansible -i inv all -m shell -a "curl 10.100.237.21 -s"
10.0.150.224 | success | rc=0 >>
404 page not found

10.0.124.137 | success | rc=0 >>
404 page not found

10.0.162.129 | FAILED | rc=56 >>
```

Same test here with scale cmd=1
```
ubuntu@ip-10-0-8-122:/tmp$ ansible -i inv controller -m shell -a "./kubectl --namespace=alfred-kube get pods -o wide"
10.0.124.137 | success | rc=0 >>
NAME                       READY     STATUS    RESTARTS   AGE       NODE
alfred-kube-v2-cmd-d1uvz   1/1       Running   0          15h       10.0.150.224

ubuntu@ip-10-0-8-122:/tmp$ ansible -i inv all -m shell -a "curl 10.100.237.21 -s"
10.0.150.224 | success | rc=0 >>
404 page not found

10.0.124.137 | FAILED | rc=56 >>


10.0.162.129 | FAILED | rc=56 >>
```

I actually checked the router mesh configuration and it is correct, it send the traffic to the correct VIP, but the problem seems to be that the the VIP is not known on certain host. 
As a result, since the ELB sends traffic to all nodes, part of my requests return with a  5xx error. 

My setup is rather simple, so I assumed I am doing something wrong, is it a normal behaviour ? or am I missing something ? I would really appreciate a hint. 

Thanks

**Comment by aledbf** [comment]
@Grindizer if the `Host` header is missing your traffic will reach a default error [page](https://github.com/deis/deis/blob/master/router/rootfs/etc/confd/templates/nginx.conf#L373)
Please try something like `curl 10.100.237.21 -s -H 'Host: your-app-name.your-domain.com'`

**Comment by Grindizer** [comment]
hi @aledbf 
```
ubuntu@ip-10-0-8-122:~$ ansible -i inv all -m shell -a "curl 10.100.237.21 -s -H 'Host: alfred-kube.k8s-deis-1.hdmtech.net'"
10.0.150.224 | success | rc=0 >>
404 page not found

10.0.162.129 | FAILED | rc=56 >>


10.0.124.137 | FAILED | rc=56 >>
```

Same here, as I understand, the host name is need at the router mesh level to redirect to the right VIP, 
in my test I am curl-ing the VIP directly, so no need for the Host there. 

By the way this is what I have on /opt/nginx/conf/nginx.conf on one of the router-mesh container

```

    upstream alfred-kube {

        server 10.100.237.21:80;
    }


    ## server entries for custom domains

    ## end entries for custom domains

    server {
        server_name ~^alfred-kube\.(?<domain>.+)$;
        include deis.conf;

        location / {

            proxy_buffering             off;
            proxy_set_header            Host $host;
            set $access_ssl 'off';
            set $access_port '80';
            if ($access_scheme ~ https) {
                set $access_ssl 'on';
                set $access_port '443';
            }
            proxy_set_header            X-Forwarded-Port  $access_port;
            proxy_set_header            X-Forwarded-Proto $access_scheme;

            proxy_set_header            X-Forwarded-For   $proxy_protocol_addr;

            proxy_set_header            X-Forwarded-Ssl   $access_ssl;
            proxy_redirect              off;
            proxy_connect_timeout       30s;
            proxy_send_timeout          1300s;
            proxy_read_timeout          1300s;
            proxy_http_version          1.1;
            proxy_set_header            Upgrade           $http_upgrade;
            proxy_set_header            Connection        $connection_upgrade;

            proxy_next_upstream         error timeout http_502 http_503 http_504;





            proxy_pass                  http://alfred-kube;
        }


    }
```

The router is redirecting to the VIP that Kubernetes has created for the service.

**Comment by aledbf** [comment]
```
10.0.150.224 | success | rc=0 >>
404 page not found

10.0.162.129 | FAILED | rc=56 >>

10.0.124.137 | FAILED | rc=56 >>
```

that output indicates that you could have some routing issues and that `alfred-kube` pod is running in 10.0.150.224. Please run:
```
kubectl -s http://<k8s api server ip>:8080 get pods|grep alfred-kube
kubectl -s http://<k8s api server ip>:8080 describe pods alfred-kube-<missing v and type> | grep IP
```

or from 10.0.150.224 just run `docker ps| grep alfred-kube` and `docker inspect <docker name> |grep IPAddress

to get the ip address of the running container running your application.
From other node you can test if there's a routing issue running `curl -s <container ip>`

**Comment by Grindizer** [comment]
here:
```
ubuntu@ip-10-0-8-122:~$ ansible -i inv controller -m shell -a "./kubectl --namespace=alfred-kube get pods | grep alfred-kube"
10.0.124.137 | success | rc=0 >>
alfred-kube-v2-cmd-d1uvz   1/1       Running   0          17h

ubuntu@ip-10-0-8-122:~$ ansible -i inv controller -m shell -a "./kubectl --namespace=alfred-kube describe pods alfred-kube-v2-cmd-d1uvz"
10.0.124.137 | success | rc=0 >>
Name:				alfred-kube-v2-cmd-d1uvz
Namespace:			alfred-kube
Image(s):			10.0.124.137:5000/alfred-kube:v2
Node:				10.0.150.224/10.0.150.224
Labels:				name=alfred-kube,type=cmd,version=v2
Status:				Running
Reason:
Message:
IP:				172.17.0.7
Replication Controllers:	alfred-kube-v2-cmd (1/1 replicas created)
Containers:
  alfred-kube-cmd:
    Image:		10.0.124.137:5000/alfred-kube:v2
    State:		Running
      Started:		Mon, 09 Nov 2015 23:42:03 +0000
    Ready:		True
    Restart Count:	0
Conditions:
  Type		Status
  Ready 	True
No events.
```

On 10.0.150.224 I have:

```
core@ip-10-0-150-224 ~ $ docker inspect e905cc3bf2ff| grep IPAddress
        "IPAddress": "",
        "SecondaryIPAddresses": null,
```

but if I grep on the virtual IP I can see it injected into container environment:
```
core@ip-10-0-150-224 ~ $ docker inspect e905cc3bf2ff| grep 10.100.237.21
            "APP_15746_SERVICE_HOST=10.100.237.21",
            "APP_15746_PORT=tcp://10.100.237.21:80",
            "APP_15746_PORT_80_TCP=tcp://10.100.237.21:80",
            "APP_15746_PORT_80_TCP_ADDR=10.100.237.21",
```

(Thank you for your help by the way)

**Comment by aledbf** [comment]
@Grindizer from the kubectl output the ip is not in the correct subnet. I think this error is related to [this]( https://github.com/deis/deis/pull/4721) issue (flannel is not started) so each node uses the default docker subnet and because of this there's no overlay network. Because of this the VIPs will work only in 1 node.
The fix is really simple, just add the start command and reboot the nodes.

**Comment by mboersma** [comment]
> The fix is really simple, just add the start command and reboot the nodes

This fix for the k8s scheduler tech preview is in [Deis v1.12.1](https://github.com/deis/deis/releases/tag/v1.12.1), or as @aledbf pointed out, you could manually make [these changes](https://github.com/deis/deis/commit/b07db69e9b3ad197d869fec4cb0b5a7051a0ef16) to `flanneld.service` and restart it.

**Comment by Grindizer** [comment]
Thanks, I actually tried the fix on 1.11 but could not make it work, probably missed something else, I will move to 1.12 anyway.

Thank you guys

**Comment by krancour** [comment]
@Grindizer it's also possible you were affected by the problem that is fixed by #4742

**Comment by bacongobbler** [comment]
closing in favour of #4860. The tech previews are being removed because we're finished with the experiment and these were never meant to be used in production.

If users want to play around with Deis on kubernetes, please feel free to follow the documentation at https://github.com/deis/workflow#hacking-workflow. Thanks!
